File: apps/mobile/lib/engine/animator.dart
import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';
import 'types.dart';

typedef StrokeFrameCallback = void Function(List<Stroke> partial, double progress);

class StrokeAnimator {
  final List<Stroke> strokes;
  StrokeAnimator(this.strokes);

  bool _running = false;
  int? _frameId;
  double _startTime = 0;

  bool get isPlaying => _running;

  void play({
    double speedMultiplier = 2,
    int delayBeforeStart = 200,
    int delayAfterEnd = 500,
    StrokeFrameCallback? onFrame,
    VoidCallback? onComplete,
    void Function(double)? onProgress,
  }) {
    if (strokes.isEmpty || _running) return;
    _running = true;

    final durations = strokes.map((s) {
      double dist = 0;
      for (int i = 1; i < s.points.length; i++) {
        dist += sqrt((s.points[i].x - s.points[i - 1].x) * (s.points[i].x - s.points[i - 1].x) +
            (s.points[i].y - s.points[i - 1].y) * (s.points[i].y - s.points[i - 1].y));
      }
      return min(3000.0, 50 + dist * 1.5);
    }).toList();

    final totalDuration = durations.reduce((a, b) => a + b) + 500;

    Future.delayed(Duration(milliseconds: delayBeforeStart), () {
      _startTime = DateTime.now().millisecondsSinceEpoch.toDouble();

      void tick() {
        final elapsed = ((DateTime.now().millisecondsSinceEpoch - _startTime) * speedMultiplier);
        final progress = min(1.0, elapsed / totalDuration);
        onProgress?.call(progress);

        final partial = <Stroke>[];
        double accumulated = 0;

        for (int i = 0; i < strokes.length; i++) {
          if (accumulated + durations[i] > elapsed) {
            final frac = max(0.0, (elapsed - accumulated) / durations[i]);
            final visibleCount = max(1, (strokes[i].points.length * frac).floor());
            partial.add(Stroke(
              id: strokes[i].id,
              points: strokes[i].points.take(visibleCount).toList(),
              tool: strokes[i].tool,
              brush: strokes[i].brush,
              stickerId: strokes[i].stickerId,
              timestamp: strokes[i].timestamp,
            ));
            break;
          } else {
            partial.add(strokes[i]);
            accumulated += durations[i];
          }
        }

        onFrame?.call(partial, progress);

        if (progress < 1) {
          _frameId = null;
          WidgetsBinding.instance.addPostFrameCallback((_) => tick());
        } else {
          _running = false;
          Future.delayed(Duration(milliseconds: delayAfterEnd), () => onComplete?.call());
        }
      }

      WidgetsBinding.instance.addPostFrameCallback((_) => tick());
    });
  }

  void stop() {
    _running = false;
  }
}

File: engine/src/core/animator.ts
import type { Stroke, Point } from "../types";

/** Replay strokes chronologically for a satisfying playback animation */
export interface AnimationConfig {
  speedMultiplier: number; // 1 = realtime, 2 = 2x, 0.5 = half-speed
  delayBeforeStart: number; // ms
  delayAfterEnd: number; // ms
  onProgress?: (progress: number) => void; // 0-1
  onFrame?: (partialStrokes: Stroke[], currentTime: number) => void;
  onComplete?: () => void;
}

export class StrokeAnimator {
  private _frameId: number | null = null;
  private _startTime: number = 0;
  private _running: boolean = false;

  constructor(private strokes: Stroke[]) {}

  /** Start playing back strokes */
  play(config: AnimationConfig): void {
    if (this.strokes.length === 0 || this._running) return;
    this._running = true;

    const t0 = this.strokes[0].timestamp;
    const tLast = this.strokes[this.strokes.length - 1].timestamp;
    const total = Math.max(1, tLast - t0 + 2000); // add tail buffer

    // Pre-calculate cumulative durations for each stroke
    const durations = this.strokes.map((s) => {
      const dist = s.points.reduce(
        (sum, p, i) => (i === 0 ? 0 : sum + Math.hypot(p.x - s.points[i - 1].x, p.y - s.points[i - 1].y)),
        0
      );
      return Math.min(3000, 50 + dist * 1.5); // min 50ms, max 3000ms
    });
    const totalDuration = durations.reduce((s, d) => s + d, 0) + 500;

    setTimeout(() => {
      this._startTime = performance.now();

      const tick = () => {
        const elapsed = (performance.now() - this._startTime) * config.speedMultiplier;
        const progress = Math.min(1, elapsed / totalDuration);
        config.onProgress?.(progress);

        // Build partial strokes for this frame
        const partial: Stroke[] = [];
        let accumulated = 0;

        for (let i = 0; i < this.strokes.length; i++) {
          if (accumulated + durations[i] > elapsed) {
            // Partial reveal of this stroke
            const frac = Math.max(0, (elapsed - accumulated) / durations[i]);
            const visibleCount = Math.max(1, Math.floor(this.strokes[i].points.length * frac));
            partial.push({
              ...this.strokes[i],
              points: this.strokes[i].points.slice(0, visibleCount),
            });
            break;
          } else {
            // Full stroke visible
            partial.push(this.strokes[i]);
            accumulated += durations[i];
          }
        }

        config.onFrame?.(partial, elapsed);

        if (progress < 1) {
          this._frameId = requestAnimationFrame(tick);
        } else {
          this._running = false;
          this._frameId = null;
          setTimeout(() => config.onComplete?.(), config.delayAfterEnd);
        }
      };

      this._frameId = requestAnimationFrame(tick);
    }, config.delayBeforeStart);
  }

  stop(): void {
    this._running = false;
    if (this._frameId !== null) {
      cancelAnimationFrame(this._frameId);
      this._frameId = null;
    }
  }

  get isPlaying(): boolean {
    return this._running;
  }

  /** Serialize all strokes as a replayable animation sequence */
  static serialize(strokes: Stroke[]): string {
    return JSON.stringify(
      strokes.map((s) => ({
        tool: s.tool,
        brush: s.brush,
        points: s.points,
        stickerId: s.stickerId,
        duration: s.points.length > 1
          ? s.points.reduce(
              (sum, p, i) => (i === 0 ? 0 : sum + Math.hypot(p.x - s.points[i - 1].x, p.y - s.points[i - 1].y)),
              0
            ) * 1.5
          : 500,
      })),
      null,
      2
    );
  }
}
